### PR TITLE
Construct C++ `NaryOpNode` from `Array*`

### PR DESCRIPTION
**Pull request by arcondello**

### CODE REVIEW

File: dwave/optimization/src/nodes/mathematical.cpp [hunk 1]
```diff
@@ -269,12 +269,28 @@ NaryOpNode<BinaryOp>::NaryOpNode(Node* node_ptr)
...
+    for (Node* ptr : node_ptrs) {
+        add_node(ptr);
+    }
+}
+
+template <class BinaryOp>
+NaryOpNode<BinaryOp>::NaryOpNode(std::span<Array*> array_ptrs)
+        : Node(), ArrayOutputMixin(nonempty(array_ptrs)->shape()) {
+    for (Array* ptr : array_ptrs) {
+        add_node(dynamic_cast<Node*>(ptr));
```
Comment (alexzucca90) [inline dwave/optimization/src/nodes/mathematical.cpp:293]: Should we add an assertion that the casting doesn't return `nullptr`?

Comment (arcondello) [reply dwave/optimization/src/nodes/mathematical.cpp:293]: We should. Also, turns out `add_node()` casts to `Array*` .... Will clean that up too.

Comment (arcondello) [reply dwave/optimization/src/nodes/mathematical.cpp:297]: Ok, added some checks. We still do two dynamic casts, but IMO the right fix is deeper. So we can/should live with it for now.

**Review by alexzucca90 — APPROVED** [review]
lgtm

**Pull request merged**